### ISSUE TITLE
[Done] Question Regarding Missing Model Files in Open-Reasoner-Zero / orz_ckpt Folder

### ISSUE THREAD

**Issue by junkangwu** [Open-Reasoner-Zero/Open-Reasoner-Zero#22]
Hi, I am reproducing your experiment with 4 servers and 8 A100 GPUs. So far, the training curve looks very normal, but I noticed an output in the log file:

```
Episode [1/20]:  11%|█         | 50/445 [6:55:57<58:14:56, 530.88s/it]2025-02-21 22:59:11.882 | INFO     | orz.ppo.trainer:train:182 - Successfully save model weights, training continue.
```

However, in the `Open-Reasoner-Zero/orz_ckpt` folder, there are no files other than the JSON file. Why is that?

**Comment by junkangwu** [comment]
Btw, I would also like to ask about the meanings of the training parameters:

```python
num_episodes: int = 20
rollout_batch_size: int = 128 if not DEBUG_MODE else 16
n_samples_per_prompt: int = 64 if not DEBUG_MODE else 2
micro_rollout_batch_size: int = 128 if not DEBUG_MODE else 128

policy_update_steps: int = 1
critic_update_steps: int = 12 if not DEBUG_MODE else 1
micro_train_batch_size: int = 1
```

My understanding is as follows:

- **num_episodes**: I believe this represents the number of training iterations, similar to the concept of epochs?
- **rollout_batch_size**: This is the number of prompts used for rollout when processing 57k prompts, which are then used for training.
- **n_samples_per_prompt**: This indicates how many responses are generated per prompt using vllm. If this is GRPO, it will be used for baseline estimation.
- **micro_rollout_batch_size**: I noticed that this parameter is not used in the ORZ codebase.
- **policy_update_steps** and **critic_update_steps**: These represent the number of steps for updating the policy model and critic model within the current batch, i.e., the number of gradient updates.
- **micro_train_batch_size**: Is this the number of sampling iterations for the data? Does this refer to how many data points are sampled from the DataLoader at a time?

I hope to receive your help. Is my understanding correct, or are there any misconceptions? Your response would greatly enhance my understanding of your code. Thank you!

**Comment by junkangwu** [comment]
When I run the experiments, I notice that the updates for the actor sometimes display:

```python
Actor Train epoch [1/1]:  91%|█████████ | 10/11 [03:10<00:15, 15.08s/it, pg=0.0177, ret=0.0026, glen=450, tlen=652, kl=0, act_lr=2e-8, ent=0.498]
Actor Train epoch [1/1]:  91%|█████████ | 10/11 [03:10<00:19, 19.09s/it, pg=0.0177, ret=0.0026, glen=450, tlen=652, kl=0, act_lr=2e-8, ent=0.498]
```

And at other times, it displays:

```python
Actor Train epoch [1/1]:  94%|█████████▍| 17/18 [02:06<00:06,  6.57s/it, pg=0.033, ret=0.0383, glen=852, tlen=1.07e+3, kl=0.0827, act_lr=1e-6, ent=0.0525]  
Actor Train epoch [1/1]:  94%|█████████▍| 17/18 [02:06<00:07,  7.41s/it, pg=0.033, ret=0.0383, glen=852, tlen=1.07e+3, kl=0.0827, act_lr=1e-6, ent=0.0525]
```

What factors influence the current number of steps for the actor training being either 11 or 18? How is this value estimated?

**Comment by Hanqer** [comment]
Hi, @junkangwu, 

For your first question, it is wired. I will try to reproduce this.

The second question:
* For num_episodes, rollout_batch_size, n_samples_per_prompt , your understanding is exactly correct. 
* micro_rollout_batch_size, it is used in ppo_base_exp.py, which means the max number of sequences handled by per vllm dp rank in a time. 
* update_steps, these mean the parameter update steps (same to optimizer step) in one generation batch, not the gradient update steps, since the gradient was accumulated multiple times in backwards. 
* micro_train_batch_size, it should always be 1 in current implementation. It is parsed to deepspeed to control the accumulations.

The third question: number of steps is actually the forward/backward times of action and critic models. It is a dynamic value because of the total tokens generated by vllms is not a constant value, and the packing process has a fixed max seqlen.

Hope it will helpful to you. If you have any question, please let me know.

**Comment by junkangwu** [comment]
Thank you for your quick and patient response. I really appreciate it!

**Comment by junkangwu** [comment]
I have another question:

```python
policy_update_steps: int = 1
critic_update_steps: int = 12 if not DEBUG_MODE else 1
```

This means that the policy model is updated 12 times before the critic model is updated once. What is the purpose of this setting? Is it based on experience?

@Hanqer

**Comment by REIGN12** [comment]
> I have another question:
> 
> policy_update_steps: int = 1
> critic_update_steps: int = 12 if not DEBUG_MODE else 1
> This means that the policy model is updated 12 times before the critic model is updated once. What is the purpose of this setting? Is it based on experience?
> 
> [@Hanqer](https://github.com/Hanqer)

This is specially designed and we illustrate the motivation in our paper~
Basically means for each rollout step, policy will update parameters once, while critic will update for at most 12 times.
<img width="996" alt="Image" src="https://github.com/user-attachments/assets/0236f4bf-a16a-4b49-afa5-e1820828bf99" />

**Comment by junkangwu** [comment]
Thanks for your explanation.

By the way, the issue of missing model files in the Open-Reasoner-Zero / orz_ckpt folder exists during multi-node training. In other words, when I train on 8x A100 GPUs, it works normally.

**Comment by LinglingGreat** [comment]
same issue, even when I train on 8x A100 GPUs. I didn't use Hydra by the way.

**Comment by LinglingGreat** [comment]
I found it. It's stored in my home directory, not in the directory where my code is located.
